bdai_ros2_wrappers/bdai_ros2_wrappers/process.py

```diff
@@ -0,0 +1,217 @@
+        return self._executor
+
+
+MainCallableWithArgs = typing.Callable[[typing.Optional[typing.Sequence[str]]], typing.Optional[int]]
+MainCallableWithNoArgs = typing.Callable[[], typing.Optional[int]]
+MainCallable = typing.Union[MainCallableWithNoArgs, MainCallableWithArgs]
+
+
+class ROSAwareProcess:
+    """
+    A ``main``-like function wrapper that binds a ROS aware scope to each function invocation.
+
+    Only one process can be invoked at a time, becoming the ``current`` process.
+    This enables global access to the process and, in consequence, to its scope.
+
+    This wrapper will rarely be used explicitly but through the `wrap` decorator.
+    """
+
+    lock = threading.Lock()
+
+    current: typing.Optional["ROSAwareProcess"] = None
+
+    def __init__(self, func: MainCallable, name: typing.Optional[str] = None, **kwargs: typing.Any):
+        """
+        Initializes the ROS aware process.
+
+        Args:
+            func: a ``main``-like function to wrap.
+            name: an optional name for this process and the underlying node.
+            If not provided, the basename without extension of the calling executable will be used.
+
+        Additional keyword arguments will be forwarded to the process' `ROSAwareScope` instance.
+        """
+        self._func = func
+        if name is None:
+            name = pathlib.Path(sys.argv[0]).stem
+        self._name = name
+        self._kwargs = kwargs
+        self._scope: typing.Optional[ROSAwareScope] = None
+        functools.update_wrapper(self, self._func)
+
+    def __getattr__(self, name: str) -> typing.Any:
+        """Gets attribute from scope, if any."""
+        if self._scope is None:
+            return super().__getattribute__(name)
+        return getattr(self._scope, name)
+
+    def __call__(self, args: typing.Optional[typing.Sequence[str]] = None) -> typing.Optional[int]:
+        """
+        Invokes wrapped function in a ROS aware scope.
+
+        Args:
+            args: optional command line-like arguments.
+
+        Returns:
+            an optional return code.
+
+        Raises:
+            RuntimeError: if a ROS aware process is already running.
+        """
+        if not ROSAwareProcess.lock.acquire(blocking=False):
+            raise RuntimeError("Process already running!")
+        try:
+            rclpy.init(args=args)
+            try:
+                with ROSAwareScope(self._name, **self._kwargs) as scope:
+                    self._scope = scope
+                    ROSAwareProcess.current = self
+                    try:
+                        sig = inspect.signature(self._func)
+                        if not sig.parameters:
+                            func_no_args = typing.cast(MainCallableWithNoArgs, self._func)
+                            return func_no_args()
+                        func_with = typing.cast(MainCallableWithArgs, self._func)
+                        return func_with(rclpy.utilities.remove_ros_args(args))
```

Comment:
If this function throws an exception or error, is there a way to easily recover the thread or will it just exit? Asking because I see a case where something can die ungracefully in this thread with something else outside the scope depending on it (if I am reading this correctly).

Reply:
First of all, bear in mind that this is designed to decorate top level functions e.g. `main`. If an exception manages to propagate all the way up there, there's nothing beyond but the interpreter. 

It could happen that the thread that "owns" the node dies while another thread that "borrows" it remains alive, but I think all of those can be labelled as API misuse.